### ISSUE TITLE
Inquiry about file save function and dump function

### ISSUE THREAD

**Issue by daeinlee** [axhlzy/Il2CppHookScripts#29]
hello. Thank you for creating a really great program.
The game I'm checking has many classes and functions, so if I use functions like "c() == list_classes", wouldn't it be possible to save the list locally as a file?

Also, if you look at the wiki, you don't see the dump function. Has it disappeared now?
The game I'm checking generates an error when using il2cppdumper, so I'm asking if I can use the dump function.

**Comment by daeinlee** [comment]
ReferenceError: 'showAsm' is not defined!
Did I do something wrong?

![showasM](https://github.com/axhlzy/Il2CppHookScripts/assets/34927021/42306f51-0bef-45fd-be49-006de4fc3b60)

**Comment by axhlzy** [comment]
I don't quite understand what you're trying to express. I just retested using 'showAsm,' and it seemed quite normal.
<img width="881" alt="1699881302146" src="https://github.com/axhlzy/Il2CppHookScripts/assets/20512058/2df8d2b6-eeef-4350-82fe-960bf44b9061">


dump function ？ You means "Il2Cpp.dump()" ?
<img width="497" alt="1699881401962" src="https://github.com/axhlzy/Il2CppHookScripts/assets/20512058/6fb1aa2e-5070-4471-a3fe-1e0b2df59023">
<img width="743" alt="1699881409777" src="https://github.com/axhlzy/Il2CppHookScripts/assets/20512058/9c88bddb-1ff6-4206-bbe1-465db5a5edf0">
It seems like there are no issues with this function call either.

As for the issue with il2cppdumper not producing output, you can manually search for 'codeRegistration' and 'metadataRegistration.'

**Comment by daeinlee** [comment]
> metadataRegistration



> I don't quite understand what you're trying to express. I just retested using 'showAsm,' and it seemed quite normal. <img alt="1699881302146" width="881" src="https://user-images.githubusercontent.com/20512058/282470408-2df8d2b6-eeef-4350-82fe-960bf44b9061.png">
> 
> dump function ？ You means "Il2Cpp.dump()" ? <img alt="1699881401962" width="497" src="https://user-images.githubusercontent.com/20512058/282470829-6fb1aa2e-5070-4471-a3fe-1e0b2df59023.png"> <img alt="1699881409777" width="743" src="https://user-images.githubusercontent.com/20512058/282470834-9c88bddb-1ff6-4206-bbe1-465db5a5edf0.png"> It seems like there are no issues with this function call either.
> 
> As for the issue with il2cppdumper not producing output, you can manually search for 'codeRegistration' and 'metadataRegistration.'

답변감사합니다.
제가 "frida -U -f com.android.xxx _Ufunc.js" 스크립트를 로드해서 사용해서 그런 것으로 의심됩니다. 
이렇게 사용해도 i(), c(), m() 등이 정상적으로 사용되서, 이 상태가 정상인 것으로 착각한 것 같습니다.

내일 "fat -r com.android.xxx" 이 방식으로 해보겠습니다. 
만약, 덤프를 한다고 가정하면 아래와 같이 사용하면 될까?
1. "far -r com.android.xxx"
2. "showAsm(0x0000,0x50)

덤프의 경우는 wiki>Contents>dumpSo>/dumpMem : dump So / dump mem 링크를 누르면, 다시 화면의 최상단으로 이동해서 물어봤습니다.
찾고자하는 내용은 il2cppdump가 맞습니다. 
이것도 다시 해보겠습니다.

친절히 설명해 주셔서 감사합니다.

**Comment by daeinlee** [comment]
Thank you for your kind response.

I apologize. I realized that I had not fully understood certain parts. My questions also seemed a bit unorganized, so I have arranged them below.

**1. "showAsm" Issue**
I misunderstood how to use it before. After trying again, I confirmed that "showAsm" works well. Thank you.

**2. Query about dump**
On the wiki (https://github.com/axhlzy/Il2CppHookScripts/wiki/API), clicking on the "dumpSo / dumpMem: dump So / dump mem" item does not lead to a description of dump. Instead, it redirects to the top of the page. Therefore, I was unable to obtain information about dump and am asking about it.

**3. Query about Il2Cpp.dump**
After seeing the example picture you provided, I tried using "Il2Cpp.dump()" but encountered an error as shown in the picture below. Can you help me understand what this error is? (It dumps normally in other games.) As additional information, il2CPPDUMPER also did not dump due to an error, and other dumpers also experienced errors.
- metadata 29 version / unity 2021.3.22f1
![il2cppdump error](https://github.com/axhlzy/Il2CppHookScripts/assets/34927021/df3dba27-4ba5-4681-be0f-6c87ede4517f)


**4. Saving Function Information as an Alternative to Il2Cpp.dump**
This part corresponds to the question I asked earlier about "wouldn't it be possible to save the list locally as a file?"

Although Il2Cpp.dumps results in an error, the class and function information obtained using i()/c()/m() seem accurate. Therefore, I plan to save the function information as a file using i()/c()/m() instead of dump. (It might take a long time.) If you have time, I would really appreciate any tips you could provide.

Since my knowledge of programming is limited, there may be some parts of my explanation that are hard to understand. Please consider this. Once again, I thank you for creating such a useful program.

**Comment by axhlzy** [comment]
2. Query about dump
你提到的这个我已经把超链接添加上去了
（ps:有一些比较直观简单的我就没有截图展示了，就直接去源码中搜 globalthis.xxx去简单查看一下如何使用就好了）
3/4
用不了的话可能是游戏有加密处理或者是什么的，或者是这个项目使用到的frida-il2cpp-bridge的版本太低了？可以去frida-il2cpp-bridge它的库下载它最新的源码进行编译得到的js，附加上去之后再使用Il2Cpp.dump。单纯说保存日志，还有一个比较笨的方法，使用frida -o 重定向输出日志

![1699932432464](https://github.com/axhlzy/Il2CppHookScripts/assets/20512058/bc5079fc-01ec-4164-b5e9-00b4582c3549)

**Comment by axhlzy** [comment]
-> ***2.Query about dump***
"You mentioned this, and I have already added the hyperlink. (PS: For some more straightforward and simple ones, I didn't show screenshots. Just go directly to the source code and search for globalthis.xxx to quickly check how to use it.)

-> ***3. Query about Il2Cpp.dump | 4. Saving Function Information as an Alternative to Il2Cpp.dump***
If it doesn't work, it might be because the game has encryption processes, or maybe the version of frida-il2cpp-bridge used in this project is too old? You can go to the frida-il2cpp-bridge repository, download its latest source code, compile it to get the JS file, attach it, and then use Il2Cpp.dump. As for simply saving logs, there's another more basic method: use `frida -o` to redirect the output of the logs."
![1699932432464](https://github.com/axhlzy/Il2CppHookScripts/assets/20512058/dbf8d3d9-364a-4362-b2d3-9f01e347b94b)

**Comment by daeinlee** [comment]
"Ah! Thank you very much for your kind response.
I will try to make good use of it.

I have already tried using frida-il2cpp-bridge, but the same issue was occurring.
It seems like I will have to complete it by logging during the time I spend searching for the tool."

**Comment by daeinlee** [comment]
I'm really grateful for your help, "axhlzy". 
Thanks to you, I was able to obtain the entire list of functions. 

However, my goal was to acquire the symbols of all functions in the il2cpp segment, but I only managed to obtain a portion of them (about a quarter). 
The symbols I did acquire closely match the list obtained through ZygiskDump, confirming that l2cppHooker can be a useful tool for dealing with situations where il2cppdumper does not work (although I haven't verified this with other projects, I suspect it will have a similar effect).

I would like to express my sincere gratitude to the developer for creating such a useful tool. 
I'm interested in other useful features and plan to use them as wel
